[CPU] Improve CPU Docker build

maryamtahhan opened the pull request:
## Purpose
This PR improves the CPU Docker build process for better compatibility with Podman and adds build configuration tracking through OCI labels.

**Changes:**
1. **Podman Compatibility**: Replace bind mounts with COPY for requirements files in `docker/Dockerfile.cpu` to improve compatibility with Podman
2. **OCI Labels**: Add standard OCI image labels and custom build configuration labels to track CPU instruction set flags (AVX2, AVX512, etc.) in built images
3. **Cross-platform Build Support**: Add support to build ARM images on x86 platforms and x86 images on ARM platforms, as well as support for building on x86 platforms without specific instruction sets. 
4. **Update CPU container build Documentation**

**Fixes:** [Resolves build failures on macOS with Podman due to bind mount incompatibility](https://github.com/vllm-project/vllm/issues/29983)

## Test Plan
## Test Plan

### 1. Test Podman Build on macOS (ARM64)
```bash
# Ensure Podman machine has sufficient memory (12GB+)
podman machine set --memory 12288 podman-machine-default

# Build ARM64 image
podman build -f docker/Dockerfile.cpu \
    --platform=linux/arm64 \
    --build-arg max_jobs=4 \
    --tag vllm-cpu-arm64 .
2. Test x86_64 Cross-Compilation with Labels
# Build x86_64 AVX512 image from ARM64 host
podman build -f docker/Dockerfile.cpu \
    --platform=linux/amd64 \
    --build-arg VLLM_CPU_AVX512=true \
    --build-arg VLLM_CPU_AVX512BF16=true \
    --build-arg VLLM_CPU_AVX512VNNI=true \
    --build-arg max_jobs=4 \
    --tag vllm-cpu-x86-avx512 .

# Verify labels are present
podman inspect vllm-cpu-x86-avx512 | jq '.[0].Config.Labels'
skopeo inspect containers-storage:localhost/vllm-cpu-x86-avx512 | jq '.Labels'
3. Verify Docker Compatibility
# Test with Docker on Linux
docker build -f docker/Dockerfile.cpu \
    --platform=linux/arm64 \
    --tag vllm-cpu-env .
```    
   
## Test Result
- vllm-cpu image builds with podman, COPY approach works with both Docker and Podman
- 86_64 cross-compilation: AVX512 image builds correctly from ARM64 host ...
- OCI labels: All build configuration labels are correctly set and visible via podman inspect and skopeo inspect

Example label output:
```bash
podman inspect vllm-cpu-x86-avx512 | jq '.[0].Config.Labels'
{
  "ai.vllm.build.cpu-amxbf16": "false",
  "ai.vllm.build.cpu-avx2": "false",
  "ai.vllm.build.cpu-avx512": "true",
  "ai.vllm.build.cpu-avx512bf16": "true",
  "ai.vllm.build.cpu-avx512vnni": "true",
  "ai.vllm.build.cpu-disable-avx512": "false",
  "ai.vllm.build.python-version": "3.12",
  "ai.vllm.build.target-arch": "amd64",
  "io.buildah.version": "1.37.4",
  "org.opencontainers.image.description": "vLLM inference engine for CPU platforms",
  "org.opencontainers.image.ref.name": "ubuntu",
  "org.opencontainers.image.source": "https://github.com/vllm-project/vllm",
  "org.opencontainers.image.title": "vLLM CPU",
  "org.opencontainers.image.vendor": "vLLM Project",
  "org.opencontainers.image.version": "22.04"
}
```
```bash
podman inspect vllm-cpu-x86-amx | jq '.[0].Config.Labels'   
{
  "ai.vllm.build.cpu-amxbf16": "true",
  "ai.vllm.build.cpu-avx2": "false",
  "ai.vllm.build.cpu-avx512": "true",
  "ai.vllm.build.cpu-avx512bf16": "true",
  "ai.vllm.build.cpu-avx512vnni": "true",
  "ai.vllm.build.cpu-disable-avx512": "false",
  "ai.vllm.build.python-version": "3.12",
  "ai.vllm.build.target-arch": "amd64",
  "io.buildah.version": "1.37.4",
  "org.opencontainers.image.description": "vLLM inference engine for CPU platforms",
  "org.opencontainers.image.ref.name": "ubuntu",
  "org.opencontainers.image.source": "https://github.com/vllm-project/vllm",
  "org.opencontainers.image.title": "vLLM CPU",
  "org.opencontainers.image.vendor": "vLLM Project",
  "org.opencontainers.image.version": "22.04"
}
```

Before these changes, simply following the instructions from the gsg to build the container image with podman results in:

```
[1/9] STEP 18/21: ENV UV_LINK_MODE="copy"
--> 9acae792160e
[1/9] STEP 19/21: RUN --mount=type=cache,target=/root/.cache/uv --mount=type=bind,src=requirements/common.txt,target=requirements/common.txt --mount=type=bind,src=requirements/cpu.txt,target=requirements/cpu.txt uv pip install --upgrade pip && uv pip install -r requirements/cpu.txt
Using Python 3.12.12 environment at: /opt/venv
Resolved 1 package in 152ms
Audited 1 package in 0.03ms
error: File not found: requirements/cpu.txt
Error: building at STEP "RUN --mount=type=cache,target=/root/.cache/uv --mount=type=bind,src=requirements/common.txt,target=requirements/common.txt --mount=type=bind,src=requirements/cpu.txt,target=requirements/cpu.txt uv pip install --upgrade pip && uv pip install -r requirements/cpu.txt": while running runtime: exit status 2
```

Also (before this patch) building on a x86 host that didn't have a particular instruction set would fail. 


---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

maryamtahhan committed:
Add cross-compilation support for CPU backend

Enable building vLLM CPU Docker images on systems without AVX2/AVX512
instruction sets by allowing explicit target ISA specification via
build arguments.

Changes:
- Add VLLM_CPU_AVX2 and VLLM_CPU_AVX512 build arguments to Dockerfile.cpu
- Add environment variable overrides in cmake/cpu_extension.cmake to
  bypass CPU feature detection when cross-compiling
- Document new build arguments in Dockerfile header and user documentation

This allows building on any build system for deployment on target
systems with specific CPU instruction sets.

Usage examples:
  # For AVX512 targets:
  docker build -f docker/Dockerfile.cpu \
    --build-arg VLLM_CPU_AVX512=true \
    --build-arg VLLM_CPU_AVX512BF16=true \
    --build-arg VLLM_CPU_AVX512VNNI=true \
    --tag vllm-cpu-env .

  # For AVX2 targets:
  docker build -f docker/Dockerfile.cpu \
    --build-arg VLLM_CPU_AVX2=true \
    --tag vllm-cpu-env .

Note: Both "true" and "1" values are supported for consistency with
existing pipelines and workflows.

Signed-off-by: Maryam Tahhan <mtahhan@redhat.com>

maryamtahhan committed:
[Docker] Improve CPU Dockerfile compatibility and ARM build documentation

This commit improves the CPU Docker build process for better compatibility
with Podman and clearer ARM64 build instructions.

Changes to docker/Dockerfile.cpu:
- Replace bind mounts with COPY for requirements files to improve
  compatibility with Podman and other container builders
- Remove .git bind mount from build steps (not required for wheel builds)
- Add comments documenting the copied requirements files
- Maintains cache mounts for performance (uv, ccache, .deps)

Changes to docs/getting_started/installation/cpu.arm.inc.md:
- Add --platform=linux/arm64 flag to docker build command
- Add note explaining the platform flag is required for ARM64 Linux images,
  especially when building on macOS ARM systems
- Fixes "no match for platform in manifest" error on macOS

The COPY approach for small requirements files is more portable across
different container runtimes while maintaining build performance through
cache mounts for large downloads and compilation artifacts.

Signed-off-by: Maryam Tahhan <mtahhan@redhat.com>

maryamtahhan committed:
[Docker] Add OCI labels to CPU release image for build configuration tracking

Add standard OCI image labels and custom build configuration labels to the
vllm-openai (release) target in the CPU Dockerfile. These labels make it
easy to inspect deployed images and verify their build configuration.

Added labels:
- Standard OCI labels (title, description, vendor, source)
- Build configuration labels under ai.vllm.build.* namespace:
  - target-arch: Target CPU architecture (amd64/arm64)
  - cpu-disable-avx512, cpu-avx2, cpu-avx512: ISA feature flags
  - cpu-avx512bf16, cpu-avx512vnni, cpu-amxbf16: Extended ISA features
  - python-version: Python version used in the image

These labels are visible via:
  podman inspect <image> | jq '.[0].Config.Labels'
  skopeo inspect containers-storage:<image> | jq '.Labels'

This helps users and operators verify which CPU instruction sets an image
was compiled for before deploying to production systems.

Signed-off-by: Maryam Tahhan <mtahhan@redhat.com>

maryamtahhan commented on docker/Dockerfile.cpu:
```diff
@@ -100,28 +111,31 @@ ENV VLLM_CPU_AMXBF16=${VLLM_CPU_AMXBF16}
     --mount=type=cache,target=/root/.cache/ccache \
     --mount=type=cache,target=/workspace/vllm/.deps,sharing=locked \
-    --mount=type=bind,source=.git,target=.git \
     VLLM_TARGET_DEVICE=cpu python3 setup.py bdist_wheel --dist-dir=dist --py-limited-api=cp38
```
COPY . . at line 120 will copy the .git directory

bigPYJ1151 commented on docker/Dockerfile.cpu:
```diff
@@ -15,6 +15,8 @@
 # Build arguments:
 #   PYTHON_VERSION=3.13|3.12 (default)|3.11|3.10
 #   VLLM_CPU_DISABLE_AVX512=false (default)|true
+#   VLLM_CPU_AVX2=false (default)|true|1 (for cross-compilation)
```
Please also remove ```1``` to keep unified

And please help to add ```(for cross-compilation)``` to ```VLLM_CPU_AVX512BF16```, ```VLLM_CPU_AVX512VNNI```, ```VLLM_CPU_AMXBF16``` as well :)

bigPYJ1151 commented on docs/getting_started/installation/cpu.x86.inc.md:
```diff
@@ -164,21 +164,84 @@ uv pip install dist/*.whl
 # --8<-- [end:pre-built-images]
 # --8<-- [start:build-image-from-source]
 
+## Building for your target CPU
```
Perhaps we can simplify this section via: 
- Just show a build command with all build arguments
- note by default ISAs will be auto detected
- note all arguments like ```VLLM_CPU_{ISA}``` are used for cross-compilation, set ```VLLM_CPU_{ISA}=true``` will build vLLM with ```ISA```, set ```VLLM_CPU_{ISA}=false``` will depends on the auto-detection
- provide some examples for different scenarios

bigPYJ1151 reviewed (commented):
Thanks! Have some initial comments :)

Looks like we should refactor the cmake script...

maryamtahhan committed:
Address PR feedback: simplify cmake conditions and documentation

- Simplify AVX2/AVX512 cmake conditions by removing redundant boolean checks
- Add '(for cross-compilation)' labels to all CPU ISA build arguments
- Revert to COPY approach for test requirements while maintaining sed workflow
- Fix AMXBF16 default value in OCI label (true not false)
- Restructure CPU build documentation for clarity:
  - Reorganize to emphasize auto-detection as default behavior
  - Add clearer examples for native builds vs cross-compilation
  - Better explain when to use cross-compilation flags

Signed-off-by: Maryam Tahhan <mtahhan@redhat.com>

fadara01 commented on cmake/cpu_extension.cmake:
```diff
@@ -13,6 +13,8 @@ endif()
 #
 # Define environment variables for special configurations
 #
+set(ENABLE_AVX2 $ENV{VLLM_CPU_AVX2})
```
Should we do something similar for `ARM_BF16_FOUND`?

bigPYJ1151 committed:
Merge branch 'main' into hotfix-container-build
